framework/src/meshmodifiers/ElementSubdomainModifierBase.C

```diff
@@ -0,0 +1,662 @@
+    }
+  }
+
+  bnd_info.parallel_sync_side_ids();
+  bnd_info.parallel_sync_node_ids();
+}
+
+void
+ElementSubdomainModifierBase::findReinitializedElemsAndNodes(
+    const std::unordered_map<dof_id_type, std::pair<SubdomainID, SubdomainID>> & moved_elems)
+{
+  // Clear cached element reinitialization data
+  _reinitialized_elems.clear();
+  _reinitialized_nodes.clear();
+
+  for (const auto & [elem_id, subdomain] : moved_elems)
+  {
+    // Default: any element that changes subdomain is reinitialized
+    if (std::find(_subdomain_ids_to_reinitialize.begin(),
+                  _subdomain_ids_to_reinitialize.end(),
+                  Moose::ANY_BLOCK_ID) != _subdomain_ids_to_reinitialize.end())
+      _reinitialized_elems.insert(elem_id);
+    else // Reinitialize if new subdomain is in list of subdomains to be reinitialized
+    {
+      const auto & [from, to] = subdomain;
+      if (subdomainIsReinitialized(to) && _old_subdomain_reinitialized)
+        _reinitialized_elems.insert(elem_id);
+      // Only reinitialize if original subdomain is not in list of subdomains
+      else if (subdomainIsReinitialized(to) && !_old_subdomain_reinitialized &&
+               !subdomainIsReinitialized(from))
+        _reinitialized_elems.insert(elem_id);
+      else // New subdomain is not in list of subdomains
+        continue;
+    }
+
+    const auto elem = _mesh.elemPtr(elem_id);
+    for (unsigned int i = 0; i < elem->n_nodes(); ++i)
+      if (nodeIsNewlyReinitialized(elem->node_id(i)))
+        _reinitialized_nodes.insert(elem->node_id(i));
+  }
+}
+
+bool
+ElementSubdomainModifierBase::subdomainIsReinitialized(SubdomainID id) const
+{
+  // Default: any element that changes subdomain is reinitialized
+  if (std::find(_subdomain_ids_to_reinitialize.begin(),
+                _subdomain_ids_to_reinitialize.end(),
+                Moose::ANY_BLOCK_ID) != _subdomain_ids_to_reinitialize.end())
+    return true;
+
+  // Is subdomain in list of subdomains to be reinitialized
+  return std::find(_subdomain_ids_to_reinitialize.begin(),
+                   _subdomain_ids_to_reinitialize.end(),
+                   id) != _subdomain_ids_to_reinitialize.end();
+}
+
+bool
+ElementSubdomainModifierBase::nodeIsNewlyReinitialized(dof_id_type node_id) const
+{
+  // If any of the node neighbors are already reinitialized, then the node is NOT newly
+  // reinitialized.
+  for (auto neighbor_elem_id : _mesh.nodeToElemMap().at(node_id))
+    if (subdomainIsReinitialized(_mesh.elemPtr(neighbor_elem_id)->subdomain_id()))
+      return false;
+  return true;
+}
+
+void
+ElementSubdomainModifierBase::applyIC(bool displaced)
+{
+  _fe_problem.projectInitialConditionOnCustomRange(reinitializedElemRange(displaced),
+                                                   reinitializedBndNodeRange(displaced));
+
+  // Set old and older solutions on the reinitialized dofs to the reinitialized values
+  setOldAndOlderSolutions(_fe_problem.getNonlinearSystemBase(_sys.number()),
```

Comment:
let's add a mooseAssert that we only have a single NL system. even though the UO is associated with a system, I bet we would prefer all systems do it

Reply:
added